test/metabase/models/login_history_test.clj

```diff
@@ -34,51 +35,61 @@
                  (#'login-history/first-login-ever? history-1))))))))))
 
 (deftest send-email-on-first-login-from-new-device-test
-  (mt/with-temp User [{user-id :id, email :email, first-name :first_name}]
-    (let [device (str (java.util.UUID/randomUUID))]
-      (testing "send email on first login from *new* device (but not first login ever)"
-        (mt/with-fake-inbox
-          (mt/with-temp* [LoginHistory [_ {:user_id user-id, :device_id (str (java.util.UUID/randomUUID))}]
-                          LoginHistory [_ {:user_id user-id, :device_id device, :timestamp #t "2021-04-02T15:52:00-07:00[US/Pacific]"}]]
-            (is (schema= {(s/eq email)
-                          [{:from    su/Email
-                            :to      (s/eq [email])
-                            :subject (s/eq (format "We've Noticed a New Metabase Login, %s" first-name))
-                            :body    [(s/one {:type    (s/eq "text/html; charset=utf-8")
-                                              :content s/Str}
-                                             "HTML body")]}]}
-                         @mt/inbox))
-            (let [message (-> @mt/inbox (get email) first :body first :content)]
-              (testing (format "\nMessage = %s" (pr-str message))
-                (is (string? message))
-                (when (string? message)
-                  (is (str/includes? message "We've noticed a new login on your Metabase account."))
-                  (is (str/includes? message "We noticed a login on your Metabase account from a new device."))
-                  (is (str/includes? message "Browser (Chrome/Windows) - Unknown location"))
-                  (if (= (mdb/db-type) :h2)
-                    (str/includes? message "April 2 3:52 PM (GMT-07:00)")
-                    (str/includes? message "April 2 10:52 PM (GMT)")))))
+  (testing "User should get an email the first time they log in from a new device (#14313, #15603)"
+    (mt/with-temp User [{user-id :id, email :email, first-name :first_name}]
+      (let [device (str (java.util.UUID/randomUUID))]
+        (testing "send email on first login from *new* device (but not first login ever)"
+          (mt/with-fake-inbox
+            ;; mock out the IP address geocoding function so we can make sure it handles timezones like PST correctly
+            ;; (#15603)
+            (with-redefs [request.u/geocode-ip-addresses (fn [ip-addresses]
+                                                           (into {} (for [ip-address ip-addresses]
+                                                                      [ip-address
+                                                                       {:description "San Francisco, California, United States"
+                                                                        :timezone    (t/zone-id "America/Los_Angeles")}])))]
```

Comment:
The `with-redefs` here caused this to fail before the fix

Reply:
if i'm understanding correctly, you might want to return some gibberish in that timezone from the geocoding to ensure that no codepaths try to use it.

Reply:
the stuff in `metabase.models.login-history/human-readable-info` does actually use it -- that's totally ok -- the issue was that `human-readable-info` converts `:timezone` to a human-friendly string (which is what we want for the API endpoint), but the email code tried to convert the string back to a `ZoneId`, which it didn't need to do.

Reply:
We actually want a valid timezone here to test the code path that was failing -- previously it was just trying to geocode `127.0.0.1` or something which didn't result in a valid timezone, so #15603 didn't get triggered

Reply:
ah gotcha